diff --git a/go/private/extensions.bzl b/go/private/extensions.bzl
--- a/go/private/extensions.bzl
+++ b/go/private/extensions.bzl
@@ -381,11 +381,7 @@
 
         # See _download_sdk below for details on these facts.
         if hasattr(ctx, "facts"):
-            kwargs["facts"] = {
-                version: sdk_info
-                for version, sdk_info in sdks_by_version.items()
-                if version in used_versions
-            }
+            kwargs["facts"] = used_versions
         return ctx.extension_metadata(**kwargs)
     else:
         return None
@@ -424,15 +420,14 @@
         # and digest here. When using a version of Bazel that supports module extension facts, this
         # info will be persisted in the lockfile, allowing for truly airgapped builds with an
         # up-to-date lockfile and download (formerly repository) cache.
-        if version not in sdks_by_version:
+        sdks = sdks_by_version.get(version)
+        if sdks == None:
             # Lazily fetch the information about all SDKs so that we avoid the download if the facts
             # already contain all the versions we care about.
-            sdks_by_version.clear()
-            sdks_by_version.update(fetch_sdks_by_version(module_ctx))
-        if version not in sdks_by_version:
+            sdks = fetch_sdks_by_version(module_ctx).get(version)
+        if sdks == None:
             fail("go_sdk: no SDKs found for version {} requested by".format(version), download_tag)
-        used_versions[version] = True
-        sdks = sdks_by_version[version]
+        used_versions[version] = sdks
 
     go_download_sdk_rule(
         name = name,
